### PR TITLE
PHR afpacket integration

### PR DESCRIPTION
**Pull request by moosingin3space**

### CODE REVIEW

**Comment by CollinValley** [comment]
Nathan, can you rebase all your changes on top of the current PHR? That should resolve any issues before we merge this.

**Comment by moosingin3space** [comment]
Rebased (and removed Drew's change as well).

File: route-rs-afpacket/tests/integration_test.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    let mut eth_pkt = packets::EthernetFrame::encap_ipv4(ipv4_pkt);
+    eth_pkt.set_src_mac(packets::MacAddr::new([0xac, 0x67, 0x3d, 0xfa, 0xbc, 0xdd]));
+    eth_pkt.set_dest_mac(packets::MacAddr::new([0xac, 0x67, 0x3d, 0xfa, 0xbc, 0xdf]));
+
+    eth_pkt.data
+}
+
+#[test]
+#[should_panic]
+#[ignore]
```
Comment (CollinValley) [inline route-rs-afpacket/tests/integration_test.rs:59]: Leave a comment why this test can't be run by default, maybe with steps to run the test by hand.

Comment (CollinValley) [reply route-rs-afpacket/tests/integration_test.rs:59]: All the other ignored tests should follow the same pattern

File: route-rs-afpacket/src/output.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+use futures::{
+    self, ready,
+    task::{Context, Poll},
+    Future, Stream,
+};
+use route_rs_runtime::link;
+use std::pin::Pin;
+
+#[derive(Default)]
+pub struct AfPacketOutputLink {
```
Comment (CollinValley) [inline route-rs-afpacket/src/output.rs]: Need documentation before we merge. 
Please describe what this link is, and how it is useful.

File: route-rs-afpacket/src/input.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    self, ready,
+    task::{Context, Poll},
+};
+use route_rs_runtime::link;
+use std::{mem, pin::Pin};
+
+const MTU: usize = 1500;
+
+#[derive(Default)]
+pub struct AfPacketInputLink {
```
Comment (CollinValley) [inline route-rs-afpacket/src/input.rs]: Need documentation before we merge. 
Please describe what this link is, and how it is useful.

**Review by CollinValley — CHANGES_REQUESTED** [review]
The code itself is in a good state. But we need doc comments on the parts that end users are expected to see, such as the links themselves.

File: route-rs-afpacket/src/output.rs (hunk 2 above)
Comment (CollinValley) [reply route-rs-afpacket/src/output.rs]: Nit: Remove Link from the name,  accidental Hungarian.

File: afpacket/src/sockets.rs [hunk 4]
```diff
@@ -242,25 +242,6 @@ impl BoundSocket {
     }
 }
 
-impl Read for BoundSocket {
```
Comment (moosingin3space) [inline afpacket/src/sockets.rs:245]: Write down why I did this.

File: afpacket/src/tokio_sockets.rs [hunk 5]
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
         self.sock.get_mut().set_promiscuous(p)
     }
 
-    pub async fn send(&mut self, frame: &[u8]) -> io::Result<usize> {
-        self.sock.write(frame).await
+    pub fn split(self) -> (SendHalf, RecvHalf) {
```
Comment (moosingin3space) [inline afpacket/src/tokio_sockets.rs:47]: Clarify where this came from (`tokio::split`) and include the license.

File: afpacket/src/tokio_sockets.rs [hunk 6]
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
...
+        half.poll_recv(cx, &mut frame)
+    }
+}
+
+impl SendHalf {
+    pub fn is_pair_of(&self, other: &RecvHalf) -> bool {
+        Arc::ptr_eq(&self.inner, &other.inner)
+    }
+
+    pub fn poll_lock_tx<'a>(
```
Comment (moosingin3space) [inline afpacket/src/tokio_sockets.rs:270]: Expand this -- "reserving a space in the transmit queue"

File: afpacket/src/tokio_sockets.rs [hunk 7]
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
...
+        let half = &mut me.half;
+        let frame = &me.frame;
+        half.poll_send(cx, &frame)
+    }
+}
+
+unsafe impl Send for RecvHalf {}
+unsafe impl Send for SendHalf {}
+unsafe impl Sync for RecvHalf {}
+unsafe impl Sync for SendHalf {}
```
Comment (moosingin3space) [inline afpacket/src/tokio_sockets.rs:322]: Note why these are safe (because `Guard`).

File: afpacket/src/tokio_sockets.rs [hunk 8]
```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
...
+            cx.waker().wake_by_ref();
+
+            Poll::Pending
+        }
+    }
+}
+
+impl Guard<'_> {
+    fn sock_pin(&mut self) -> Pin<&mut AsyncBoundSocket> {
+        // this is safe because the Inner is stored in an Arc
```
Comment (moosingin3space) [inline afpacket/src/tokio_sockets.rs:381]: Expand

File: route-rs-afpacket/src/input.rs [hunk 9]
```diff
@@ -0,0 +1,70 @@
+use afpacket;
+use futures::{
+    self, ready,
+    task::{Context, Poll},
+};
+use route_rs_runtime::link;
+use std::{mem, pin::Pin};
+
+const MTU: usize = 1500;
```
Comment (moosingin3space) [inline route-rs-afpacket/src/input.rs:10]: This should almost certainly be configurable.

File: route-rs-afpacket/src/output.rs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+impl Unpin for StreamToChannel {}
+
+impl Future for StreamToChannel {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let me = self.get_mut();
+        loop {
+            let mut lock =
```
Comment (moosingin3space) [inline route-rs-afpacket/src/output.rs]: Needs to be outside the loop.

File: route-rs-afpacket/src/output.rs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                ready!(me.egress.poll_lock_tx(cx)).expect("AfPacketOutputLink: failed to write");
```
Comment (moosingin3space) [inline route-rs-afpacket/src/output.rs]: Clarify error message.

**Review by CollinValley — APPROVED** [review]
LGTMMD

**Pull request merged**